### ISSUE TITLE
About Dockerfile for webserver

### ISSUE THREAD

**Issue by gaonn** [SixK/coco-annotator-ng#1]
Thank you for the good repository. I have one question.
restx_monkey included in the webserver api requires python3.8, but coco-annotator:python-env's python is 3.6. Is the Dockerfile written correctly?

https://github.com/SixK/coco-annotator-ng/blob/33d59fd6777ddc7dc5f4415f91cff08458e0c26b/backend/webserver/Dockerfile#L1

https://github.com/SixK/coco-annotator-ng/blob/33d59fd6777ddc7dc5f4415f91cff08458e0c26b/backend/webserver/api/__init__.py#L2

**Comment by SixK** [comment]
In fact, I keept name of original jsbroks project docker file as I don't have any account to publish any official docker image for the coco-annotator-ng project. (It's probably not a good idea to have keept original image name)
Fact is you will have to build docker images by yourself to have a higher python version:
Here is python version once image rebuilt:
```
docker run -it jsbroks/coco-annotator:python-env bash
root@b4ca23a01312:/workspace# python3 --version
Python 3.11.7
```
You will have to first run the build_gpu.sh script using "bash ./build_gpu.sh":
```
cat ./build_gpu.sh 
docker build -f ./backend/Dockerfile . -t jsbroks/coco-annotator:python-env --no-cache
docker build . -t annotator_webclient_gpu --no-cache
docker build -f ./backend/workers/Dockerfile . -t annotator_workers_gpu --no-cache
```
If you have a look at Dockerfile (./backend/Dockerfile) that build jsbroks/coco-annotator:python-env docker image in this script, you will see that first exectued line is:  
FROM python:3.11

Once build_gpu.sh is executed, you can build other images running docker-compose file depending on your needs:
```
docker-compose.build.yml is to build cpu images --> docker-compose -f docker-compose.build.yml build
docker-compose.dev.yml is to build a development image with GPU container --> docker-compose -f docker-compose.dev.yml build
docker-compose.gpu.yml is to build production image with GPU container --> docker-compose -f docker-compose.gpu.yml build
```

Once docker images built you can run coco-annotator containers:
```
production container without AI GPU container --> docker-compose -f docker-compose.yml run
production container with AI GPU container --> docker-compose -f docker-compose.gpu.yml run
development container with AI GPU container --> docker-compose -f docker-compose.dev.yml run
```
Note that default AI GPU container is about 14Gb size, coco-annotator-ng can work without it, but you will loose DEXTR and SAM-HD support. Other containers are about 2Gb each. (but as docker reuse the same base image, it's probably near 2,5Gb disk space for all 4 non GPU containers).

All thoses scripts and docker images are mostly inherited from original coco-annotator project.

**Comment by gaonn** [comment]
Thank you for your answer. I was able to successfully start it using the following order.
However, after annotating with the polygon tool, most buttons such as save and next image no longer work. (I may need to reload the page for the feature to work properly.

1. ./build_gpu.sh
2. docker-compose -f docker-compose.dev.yml build
3. docker-compose -f docker-compose.gpu.yml up

Attach the log.
Please let me know if there are any mistakes.
[start.log](https://github.com/SixK/coco-annotator-ng/files/13638140/start.log)

**Comment by SixK** [comment]
You are mixing dev and gpu. 
Here are sequences to do :
```
    ./build_gpu.sh
    docker-compose -f docker-compose.dev.yml build
    docker-compose -f docker-compose.dev.yml up
```
IHM will be on localhost:8080
or
```
    ./build_gpu.sh
    docker-compose -f docker-compose.gpu.yml build
    docker-compose -f docker-compose.gpu.yml up
```
IHM will be on localhost:5000

Note, you don't need to play  ./build_gpu.sh again.

I don't see much problems in logs.
Just this lines --> 
annotator_webserver | server=192.168.0.15:8080//socket.io/ client=172.18.0.7:55726 socket shutdown error: [Errno 9] Bad file descriptor[2023-12-11 08:01:18 +0000] [9] [DEBUG] Closing connection. 

But I don't remember if it's really a problem.

Also note that, when clicking on "export COCO" button on Dataset tab, this generate the file but don't upload it automatically. You then have to go in "Exports" tab to get it. (This is an asynchronous Task)
On image when clicking on export button (I don't remember the right button name), this should export immediatly the coco file for the image.

In Coco Annotator, If some problems appears, it's usually more visible in navigator javascript web console.

**Comment by gaonn** [comment]
Oh, sorry, I pasted the wrong command.
I was also running the following.

     1. ./build_gpu.sh
     2. docker-compose -f docker-compose.dev.yml build
     3. docker-compose -f docker-compose.dev.yml up

I looked at the JavaScript web console, and it said that "config" was not defined. (This is line 3399 of the log.)
The part where the error occurred was a process that does not exist in the base coco-annotator. Do I need to define some config somewhere?
[192.168.0.15-1702430779717.log](https://github.com/SixK/coco-annotator-ng/files/13654823/192.168.0.15-1702430779717.log)

**Comment by gaonn** [comment]
It worked after fixing the following corrections from the error contents.
Is my understanding correct?

1. Uncomment below
https://github.com/SixK/coco-annotator-ng/blob/41aaacb2ebd88af73cbd19e3897bbd9afa1ad1e2/client/src/main.js#L64

2. Modify "emitModify" of Annotation.vue as below
https://github.com/SixK/coco-annotator-ng/blob/41aaacb2ebd88af73cbd19e3897bbd9afa1ad1e2/client/src/components/annotator/Annotation.vue#L846

```
const emitModify = () => {
  const uuid = Math.random().toString(36).replace(/[^a-z]+/g, "");
  annotation.value.paper_object = compoundPath.value.exportJSON({ asString: false, precision: 1 });
  //app.__vue_app__.config.globalProperties.$socket.emit("annotation", {
  socket.io.emit("annotation", {
    uuid: uuid,
    action: "modify",
    annotation: annotation.value
  });
};
```

**Comment by SixK** [comment]
I don't understand why I can't reproduce your error, 
but I definitely forgot to modify the line in emitModify().

Uncommenting the line "// app.config.globalProperties.$ssocketio = socketio; " should be useless and I may have made a typo error in $ssocketio instead of $socketio.

emitModify() is fixed with this commit [435db30](https://github.com/SixK/coco-annotator-ng/commit/435db305c8839a5c698d01f9946a0d871117ce0f)

Thank's a lot.

**Comment by gaonn** [comment]
Thank you very much for the fix.
It worked fine, so I'll close this issue.
